source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,213 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
   }
 }
 
-void AdminHandler::AdminPerTapSinkHandle::submitTrace(
-    TraceWrapperPtr&& trace, envoy::config::tap::v3::OutputSink::Format format) {
+PerTapSinkHandlePtr
+AdminHandler::createPerTapSinkHandle(uint64_t trace_id,
+                                     envoy::config::tap::v3::OutputSink::OutputSinkTypeCase type) {
+  UNREFERENCED_PARAMETER(trace_id);
+  using ProtoOutputSinkType = envoy::config::tap::v3::OutputSink::OutputSinkTypeCase;
+
+  /**
+   * Switching on the sink type here again after doing so in TapConfigBaseImpl constructor
+   * seems a bit strange. A possible refactor in the future could involve moving all Sinks
+   * to live where the FilePerTapSink lives, and passing in the sink to use into the AdminHandler.
+   */
+
+  // Select the sink implementation to use based on type specified in YAML request body
+  switch (type) {
+  case ProtoOutputSinkType::kStreamingAdmin:
+    return std::make_unique<AdminPerTapSinkHandle>(*this);
+  case ProtoOutputSinkType::kBufferedAdmin:
+    return std::make_unique<BufferedPerTapSinkHandle>(*this);
+  case ProtoOutputSinkType::kFilePerTap:
+  case ProtoOutputSinkType::kStreamingGrpc:
+    PANIC("not implemented");
+  case ProtoOutputSinkType::OUTPUT_SINK_TYPE_NOT_SET:
+    PANIC_DUE_TO_CORRUPT_ENUM;
+  }
+  PANIC_DUE_TO_CORRUPT_ENUM; // Required so that all paths return values
+}
```

Comment:
Instead of using a switch statement here maybe just use an if statement and assert that type is either admin or streaming admin? Shouldn't the caller have already verified this?